File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.Random;
import java.util.stream.Stream;

class DnDCharacter {
    private final Random random = new Random();

    private final int strength = ability();
    private final int dexterity = ability();
    private final int constitution = ability();
    private final int intelligence = ability();
    private final int wisdom = ability();
    private final int charisma = ability();

    int ability() {
        return Stream.generate(() -> random.nextInt(6) + 1)
                     .limit(4)
                     .sorted()
                     .skip(1)
                     .mapToInt(i -> i)
                     .sum();
    }

    int modifier(int input) {
        return Math.floorDiv(input - 10, 2);
    }

    int getStrength() {
        return strength;
    }

    int getDexterity() {
        return dexterity;
    }

    int getConstitution() {
        return constitution;
    }

    int getIntelligence() {
        return intelligence;
    }

    int getWisdom() {
        return wisdom;
    }

    int getCharisma() {
        return charisma;
    }

    int getHitpoints() {
        return 10 + modifier(getConstitution());
    }
}

File: kotlin/dnd-character/src/main/kotlin/DndCharacter.kt
import kotlin.math.floor
import kotlin.random.Random
import kotlin.random.nextInt

class DndCharacter {
    val strength: Int = ability()
    val dexterity: Int = ability()
    val constitution: Int = ability()
    val intelligence: Int = ability()
    val wisdom: Int = ability()
    val charisma: Int = ability()
    val hitpoints: Int = 10 + modifier(constitution)

    companion object {
        fun ability() = (1..4).asSequence().map { Random.nextInt(1..6) }.sortedDescending().take(3).sum()

        fun modifier(score: Int) = floor((score - 10).toDouble() / 2.0).toInt()
    }
}
